src/libmongoc/src/mongoc/mongoc-client-session.c

```diff
@@ -1531,18 +1588,32 @@ _mongoc_client_session_append_read_concern (const mongoc_client_session_t *cs,
    }
 
    if (txn_state == MONGOC_INTERNAL_TRANSACTION_STARTING) {
-      /* add transaction's read concern level unless user overrides */
-      if (txn_has_level && !user_rc_has_level) {
+      /* add transaction's read concern level unless user overrides or snapshot
+       * is enabled. */
+      if (txn_has_level && !user_rc_has_level && !is_snapshot) {
```

Comment:
Based on the spec, I think we do not want to set a snapshot read for the first operation in a transaction.

> Reads with read concern level ``snapshot`` that occur outside of transactions

This differs from causal consistency, which makes no mention of excluding transactions. The [transactions spec describes how read concern is applied](https://github.com/mongodb/specifications/blob/master/source/transactions/transactions.rst#behavior-of-the-readconcern-field) on the first command of a transaction.

To accomplish that, this could set `is_snapshot` to false if `txn_state` is anything other than `MONGOC_INTERNAL_TRANSACTION_NONE`.

This may be good to suggest a spec test in the PR for DRIVERS-1607. I think this could be tested with the unified test runner.

Reply:
Great point; I think we should be testing this in the new spec tests (commented on it [here](https://github.com/mongodb/specifications/pull/1022/files#r658883220)). I changed the definition of `is_snapshot` to be
```
is_snapshot = is_find_aggregate_distinct &&
                 txn_state == MONGOC_INTERNAL_TRANSACTION_NONE &&
                 mongoc_session_opts_get_snapshot (&cs->opts);
```

Reply:
Awesome, thanks for confirming it is tested!

Reply:
Now that there's a test for "StartTransaction fails in snapshot session", I think this is covered in spec tests.